### PR TITLE
Add new user doc to translate logical plan to physical plan

### LINKED ISSUE #7306: Library Guide: Building LogicalPlans
### Is your feature request related to a problem or challenge?

Part of https://github.com/apache/arrow-datafusion/issues/7014

If we want to have DataFusion used as the core of many new systems, we need it to be as easy as possible for someone to get their idea working on top of DataFusion.

Thanks to @tshauck we now have a basic Library Users Guide ❤️ and this ticket describes expanding it out

### Describe the solution you'd like

Fill in the content of https://arrow.apache.org/datafusion/library-user-guide/building-logical-plans.html

Example Outline
1. Building LogicalPlans directly
2. Building LogicalPlans using `LogicalPlanBuilder`
3. Translating LogicalPlan --> ExecutionPlan via physical planner

The examples directory holds a bunch more of examples: https://github.com/apache/arrow-datafusion/tree/main/datafusion-examples

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jc4x4**
https://github.com/apache/datafusion/issues/7306

## Which issue does this PR close?

Closes #7306.

## Rationale for this change
New user doc for translating logical plan to physical plan

## What changes are included in this PR?
- doc change
- sample rust code
- tokio dependency to support async `create_physical_plan`

## Are these changes tested?
Run test and pasted output into md.

## Are there any user-facing changes?
Only doc

### CODE REVIEW

File: docs/Cargo.toml [hunk 1]
```diff
@@ -33,3 +33,4 @@ workspace = true
 
 [dependencies]
 datafusion = { workspace = true }
+tokio = { workspace = true }
```
Comment (edmondop) [inline docs/Cargo.toml:36]: Can you add an empty newline here ?

Comment (jc4x4) [reply docs/Cargo.toml]: Yup. Also fixed a prettier issue in CI.

**Comment by alamb** [comment]
Thank you for this PR @jc4x4  and @edmondop  -- I plan to review / help with it early next week

File: docs/src/library_logical_plan.rs [hunk 2]
```diff
@@ -76,3 +79,32 @@ fn plan_builder_1() -> Result<()> {
 
     Ok(())
 }
+
+#[tokio::test]
```
Comment (alamb) [inline docs/src/library_logical_plan.rs:83]: This file is confusing -- i don't think it is used anymore --- I added https://github.com/apache/datafusion/pull/12081 to remove it

File: datafusion/core/src/lib.rs [hunk 3]
```diff
@@ -678,6 +678,12 @@ doc_comment::doctest!(
     library_user_guide_sql_api
 );
 
+#[cfg(doctest)]
```
Comment (alamb) [inline datafusion/core/src/lib.rs:681]: I added this block which means the examples in the library guide are now run as part of `cargo test --doc`

File: docs/source/library-user-guide/building-logical-plans.md [hunk 4]
```diff
@@ -31,44 +31,52 @@ explained in more detail in the [Query Planning and Execution Overview] section
...
-    fetch,
-)?);
-
-// create a Filter plan that evaluates `id > 500` that wraps the TableScan
-let filter_expr = col("id").gt(lit(500));
-let plan = LogicalPlan::Filter(Filter::try_new(filter_expr, Arc::new(table_scan))?);
-
-// print the plan
-println!("{}", plan.display_indent_schema());
+use datafusion::common::DataFusionError;
```
Comment (alamb) [inline docs/source/library-user-guide/building-logical-plans.md:40]: I updated these examples so they compile as standalone examples

**Review by alamb — APPROVED** [review]
Thank you very much @jc4x4  -- I hope you don't mind I pushed several commits to your branch to get the examples compiling and running.  Your PR showed there were some issues with our current testing methodology.

**Comment by jc4x4** [comment]
> Thank you very much @jc4x4 -- I hope you don't mind I pushed several commits to your branch to get the examples compiling and running. Your PR showed there were some issues with our current testing methodology.

I dont mind at all. Thanks for improving it. Let me know if I need to do anything on my end.

**Comment by alamb** [comment]
Thanks again @jc4x4  and @edmondop

**Pull request merged**